README.md
# lib-rv32

![build](https://github.com/trmckay/lib-rv32i/actions/workflows/build.yml/badge.svg)
![tests](https://github.com/trmckay/lib-rv32i/actions/workflows/test.yml/badge.svg)

## Overview

lib-rv32 is a collection of Rust libraries for emulating, learning, and assembling 32-bit RISC-V
integer ISAs.

- [lib-rv32-isa](https://crates.io/crates/lib-rv32-isa): library for ISA simulation
- [lib-rv32-mcu](https://crates.io/crates/lib-rv32-mcu): reference implemenation of an MCU used in conjunction with lib_rv32_isa
- [lib-rv32-asm](https://crates.io/crates/lib-rv32-asm): library for assembling RISC-V programs
- [lib-rv32-cli](https://crates.io/crates/lib-rv32-cli): CLI tool exposing the libraries
- [lib-rv32-wasm](https://rvwasm.trmckay.com): An webapp using the library's WASM bindings.

---

## Libraries

### ISA simulator

This library can execute instructions against any memory and register file that implements
the required primitives in the traits `lib_rv32_common::traits::{Memory, RegisterFile}`. This is to
encourage usage with whatever frontend you desire.

However, reference implementations are provided in `lib_rv32_mcu::*`. The library provides
functions to read from the memory, registers, and step a single instruction. Since, the
user decides when to call these functions, these will probably fit most use-cases.

### MCU

The MCU crate provides an implemenation of `Memory` and `RegisterFile` for use with the ISA
simulator. With this, one can fully emulate an embedded RISC-V core.

### Assembler

This crate can be used to assemble simple RISC-V assembly programs. The main functions offered
by this library are:

- `assemble_ir`: assemble an instruction `&str` to a `u32`
- `assemble_program`: assemble a program `&str` to a `Vec<u32>`
- `assemble_program_buf`: assemble a `BufRead` to a `Vec<u32>`


## CLI

### Emulator

The primary use of the emulator is tracing execution of RISC-V programs and making assertions
about their behavior. It currently only supports simple binary memory images
(not ELF binaries).

Enter assertions into a JSON file (note: all numbers are strings to allow for hex or decimal radices).

`assert.json`:
```json
{
    "registers": {
        "x0": "0x0",
        "a0": "20"
    },
    "memory": {
        "0x0000": "0x00010117"
    }
}
```

Then run:
```
lrv-cli -v ./prog.bin -s 24 -a assert.json
```

This will execute `prog.bin`, stop at the PC value 0x24, and then make the assertions from `assert.json`.

The program will trace the execution instruction-by-instruction:
```
[0000]  00010117  |  auipc  sp, 0x10           |  sp <- 0x10000 (65536);
[0004]  fe010113  |  addi   sp, sp, -32        |  sp <- 0xffe0 (65504);
[0008]  00400513  |  addi   a0, zero, 4        |  a0 <- 0x4 (4);
[000c]  00500593  |  addi   a1, zero, 5        |  a1 <- 0x5 (5);
[0010]  00000097  |  auipc  ra, 0x0            |  ra <- 0x10 (16);
[0014]  018080e7  |  jalr   ra, (24)ra         |  ra <- 0x18 (24); pc <- 0x28;
...
```

When complete, it will summarize results:
```
...
[001c]  f0028293  |  addi   t0, t0, -256       |  t0 <- 0xf00 (3840);
[0020]  00a2a023  |  sw     a0, 0(t0)          |  (word *)0x00000f00 <- 0x14 (20);

Reached stop-PC.

a0 == 20
*0x00000000 == 65815
```

### Assembler

The CLI also exposes the assembler via the command line. You can assemble the file
`program.s` to `program.bin` using

`lrv-cli -cv program.s -o program.bin`

---

## Testing

This project has a very flexible testing system.

Unit-tests are provided wherever appropriate.

Additionally, to test the whole system, test programs can be added to `mcu/tests/programs`.
A test is simply a directory containing `.c` and `.s` source files and a `test_case.json`
consisting of assertions about the state of the MCU after the program is complete.

During testing, Cargo will for each test:

1. Compile it for RISC-V
2. Spin up a new MCU
3. Program it with the generated binary
4. Run the test program for some number of cycles
5. Make assertions
6. Report succes or failure

If a test fails, it will describe the error that caused the crash or the assertion that failed
and print an object dump of the compiled test binary:

```
...
[001c]  f0028293  |  addi   t0, t0, -256       |  t0 <- 0xf00 (3840);
[0020]  00a2a023  |  sw     a0, 0(t0)          |  (word *)0x00000f00 <- 0x14 (20);
Stopping because the stop PC 0x24 was reached.


Failed test: tests/programs/mul@0x00000024: Register assertion failed: (x10=0x00000014) != 0x00000018.

prog.elf:     file format elf32-littleriscv


Disassembly of section .text.init:

00000000 <start>:
   0:   00010117                auipc   sp,0x10
   4:   fe010113                addi    sp,sp,-32 # ffe0 <__global_pointer$+0xf75c>
   8:   00400513                li      a0,4
   c:   00500593                li      a1,5
...
```

Tests are run in CI, but can be run locally provided your system has `riscv(32|64)-unknown-elf-gcc`.

# TODO

- IMAC support
- Better logging
- Asynchronous I/O
- Better WASM support/demo
- More descriptive errors/stricter assembler
- E2E testing

cli/CHANGELOG.md
# 0.2

Initial release


src/bin.rs
use std::fs;
use std::io::prelude::*;
use std::io::BufReader;

use std::path::PathBuf;

use clap::{App, Arg};
use lazy_static::lazy_static;
use log::info;
use log::{Level, LevelFilter, Metadata, Record};

use lib_rv32::{assembler::*, constants::*, exec_one, mcu::*, Assertions};

const DEFAULT_MEM_SIZE: usize = 1024 * 64;

lazy_static! {
    static ref CFG: Config = Config::new();
}

static LOGGER: Logger = Logger;

enum Mode {
    Emulator,
    Assembler,
}

struct Config {
    file: PathBuf,
    mem_size: usize,
    stop_pc: Option<u32>,
    assertions: Option<PathBuf>,
    output: Option<PathBuf>,
    mode: Mode,
}

impl Config {
    fn new() -> Self {
        let matches = App::new("lib-rv32")
            .version("0.2.0")
            .author("Trevor McKay <tm@trmckay.com>")
            .about("Emulate RISC-V")
            .arg(
                Arg::with_name("file")
                    .help("File on which to act")
                    .required(true)
                    .index(1),
            )
            .arg(
                Arg::with_name("mem")
                    .short("m")
                    .long("mem")
                    .value_name("MEM_SIZE")
                    .help("Set the size of the MCU memory (default 64 KB)")
                    .takes_value(true),
            )
            .arg(
                Arg::with_name("stop")
                    .short("s")
                    .long("--stop")
                    .value_name("STOP_PC")
                    .help("Set the program counter at which to stop emulation")
                    .takes_value(true),
            )
            .arg(
                Arg::with_name("assertions")
                    .short("a")
                    .long("--assertions")
                    .value_name("ASSERTIONS_FILE")
                    .help("A JSON formatted set of assertions")
                    .takes_value(true),
            )
            .arg(
                Arg::with_name("output")
                    .short("o")
                    .long("--output")
                    .value_name("OUTPUT_FILE")
                    .help("Out-file for binary in assembler mode, or memory dump in emulator mode")
                    .takes_value(true),
            )
            .arg(
                Arg::with_name("verbose")
                    .short("v")
                    .long("verbose")
                    .help("Enable verbose logging")
                    .takes_value(false),
            )
            .arg(
                Arg::with_name("assemble")
                    .short("c")
                    .long("--assemble")
                    .help("Launch in assembler mode")
                    .takes_value(false),
            )
            .arg(
                Arg::with_name("emulate")
                    .short("e")
                    .long("--emulate")
                    .default_value("e")
                    .help("Launch in emulator mode")
                    .takes_value(false),
            )
            .get_matches();

        let mem_size = match matches.value_of("config") {
            Some(s) => str::parse(s).unwrap(),
            None => DEFAULT_MEM_SIZE,
        };
        let stop_pc = matches.value_of("stop").map(|s| {
            u32::from_str_radix(s, 16)
                .unwrap_or_else(|_| panic!("{} is not a valid hex literal.", s))
        });
        let verbose = !matches!(matches.occurrences_of("verbose"), 0);
        let path = PathBuf::from(matches.value_of("file").unwrap());
        let assertions = matches.value_of("assertions").map(PathBuf::from);
        let output = matches.value_of("output").map(PathBuf::from);

        let mode: Mode = if matches.occurrences_of("emulate") == 1 {
            Mode::Emulator
        } else if matches.occurrences_of("assemble") == 1 {
            Mode::Assembler
        } else {
            panic!("No mode provided.");
        };
        if matches.occurrences_of("emulate") == matches.occurrences_of("assemble") {
            panic!("Cannot launch in both modes.");
        }

        if verbose {
            log::set_logger(&LOGGER)
                .map(|()| log::set_max_level(LevelFilter::Info))
                .unwrap();
        }

        Config {
            file: path,
            mem_size,
            stop_pc,
            assertions,
            mode,
            output,
        }
    }
}

struct Logger;

impl log::Log for Logger {
    fn enabled(&self, metadata: &Metadata) -> bool {
        metadata.level() <= Level::Info
    }

    fn log(&self, record: &Record) {
        if self.enabled(record.metadata()) {
            print!("{}", record.args());
        }
    }

    fn flush(&self) {}
}

fn emu() {
    let assertions = CFG.assertions.as_ref().map(|p| Assertions::load(p));

    let mut mcu: Mcu = Mcu::new(CFG.mem_size);
    mcu.mem
        .program_from_file(&CFG.file)
        .expect("Could not program MCU.");

    loop {
        exec_one(&mut mcu.pc, &mut mcu.mem, &mut mcu.rf).unwrap();
        if Some(mcu.pc) == CFG.stop_pc {
            info!("\nReached stop-PC.\n");
            break;
        }
    }

    if let Some(mut assertions) = assertions {
        assertions.assert_all(&mut mcu.mem, &mut mcu.rf);
        println!();

        for assert in assertions.register_assertions {
            if assert.2 {
                println!("{} == {}", REG_NAMES[assert.0 as usize], assert.1)
            } else {
                println!(
                    "({} = {}) != {}",
                    REG_NAMES[assert.0 as usize],
                    mcu.rf.read(assert.0).unwrap(),
                    assert.1
                );
            }
        }
        println!();

        for assert in assertions.memory_assertions {
            if assert.2 {
                println!("*0x{:08x} == {}", assert.0, assert.1)
            } else {
                println!(
                    "(*0x{:08x} = {}) != {}",
                    assert.0,
                    mcu.mem.fetch(assert.0).unwrap(),
                    assert.1
                );
            }
        }
    }
}

fn asm() {
    let file = fs::File::open(&CFG.file).unwrap();
    let mut reader = BufReader::new(file);
    let words = assemble_buf(&mut reader).unwrap();

    if let Some(path) = &CFG.output {
        let mut output = fs::File::create(&path).unwrap();
        for w in words {
            output.write_all(&w.to_le_bytes()).unwrap();
        }
    }
}

fn main() {
    match CFG.mode {
        Mode::Assembler => asm(),
        Mode::Emulator => emu(),
    }
}
